ocaml/xapi/vm_sysprep.ml

```diff
@@ -0,0 +1,249 @@
+    and is not a directory. *)
+let mkdtemp ?(dir = temp_dir) ?(perms = 0o700) prefix suffix =
+  ( match Sys.file_exists dir with
+  | true when not (Sys.is_directory dir) ->
+      internal_error "s: %s is not a directory" __FUNCTION__ dir
+  | true ->
+      ()
+  | false ->
+      Unixext.mkdir_rec dir perms
+  ) ;
+  let rec try_upto = function
+    | n when n < 0 ->
+        internal_error "%s: can't create directory %S" __FUNCTION__ dir
+    | n -> (
+        let path = Filename.concat dir (temp_name prefix suffix) in
+        try Sys.mkdir path perms ; path with Sys_error _ -> try_upto (n - 1)
+      )
+  in
+  try_upto 20
+
+(** Crteate a temporary directory, and pass its path to [f]. Once [f]
+    returns the directory is removed again *)
+let with_temp_dir ?(dir = temp_dir) ?(perms = 0o700) prefix suffix f =
+  let dir = mkdtemp ~dir ~perms prefix suffix in
+  finally (fun () -> f dir) (fun () -> Unixext.rm_rec dir)
+
+(** name of the ISO we will use for a VM; this is not a path *)
+let iso_basename ~vm_uuid =
+  let now = Ptime_clock.now () |> Ptime.to_rfc3339 in
+  Printf.sprintf "sysprep-%s-%s.iso" vm_uuid now
+
+(** Create an ISO in [SR.dir] with content [unattend]. [SR.dir] is
+    created if it not already exists. Returns the path of the ISO image *)
+let make_iso ~vm_uuid ~unattend =
+  try
+    let basename = iso_basename ~vm_uuid in
+    let iso = SR.dir // basename in
+    Xapi_stdext_unix.Unixext.mkdir_rec SR.dir 0o755 ;
+    with_temp_dir ~dir:"/var/tmp/xapi" "sysprep-" "-iso" (fun temp_dir ->
+        let path = temp_dir // "unattend.xml" in
+        Unixext.write_string_to_file path unattend ;
+        debug "%s: written to %s" __FUNCTION__ path ;
+        let args = ["-r"; "-J"; "-o"; iso; temp_dir] in
+        Forkhelpers.execute_command_get_output genisoimage args |> ignore ;
+        (iso, basename)
+    )
+  with e ->
+    let msg = Printexc.to_string e in
+    Helpers.internal_error "%s failed: %s" __FUNCTION__ msg
+
+(** create a local ISO SR when necessary and update it such that it
+    recognises any ISO we added or removed *)
+let update_sr ~__context =
+  let host = Helpers.get_localhost ~__context in
+  let hostname = Db.Host.get_hostname ~__context ~self:host in
+  let label = SR.name hostname in
+  let mib n = Int64.(n * 1024 * 1024 |> of_int) in
+  let sr =
+    match SR.find_opt ~__context ~label with
+    | Some sr ->
+        sr
+    | None ->
+        let device_config = [("location", SR.dir); ("legacy_mode", "true")] in
+        Xapi_sr.create ~__context ~host ~name_label:label ~device_config
+          ~content_type:"iso" ~_type:"iso" ~name_description:"Sysprep ISOs"
+          ~shared:false ~sm_config:[] ~physical_size:(mib 512)
+  in
+  Xapi_sr.scan ~__context ~sr ;
+  sr
+
+(** Find the VBD for the CD drive on [vm] *)
+let find_cdr_vbd ~__context ~vm =
+  let vbds = Db.VM.get_VBDs ~__context ~self:vm in
+  let vbds' =
+    List.map (fun self -> (self, Db.VBD.get_record ~__context ~self)) vbds
```

Comment:
An alternative is to use `Db.VBD.get_records_where` to avoid multiple DB calls if there are many VBDs. You can also include the filter below and use `Db.VBD.get_refs_where`.